sdk/core/azure-core-http-netty/src/main/java/com/azure/core/http/netty/implementation/DeferredHttpProxyProvider.java

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.http.netty.implementation;
+
+import com.azure.core.http.ProxyOptions;
+import com.azure.core.util.AuthorizationChallengeHandler;
+import io.netty.bootstrap.Bootstrap;
+import io.netty.channel.Channel;
+import reactor.netty.ConnectionObserver;
+import reactor.netty.NettyPipeline;
+
+import java.net.InetSocketAddress;
+import java.net.SocketAddress;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+import java.util.regex.Pattern;
+
+/**
+ * This class defers supplying a channel pipeline with a proxy handler.
+ */
+public class DeferredHttpProxyProvider implements Function<Bootstrap, BiConsumer<ConnectionObserver, Channel>> {
+    private final AuthorizationChallengeHandler challengeHandler;
+    private final AtomicReference<ChallengeHolder> proxyChallengeHolderReference;
+    private final InetSocketAddress proxyAddress;
+    private final String username;
+    private final String password;
+    private final String nonProxyHosts;
+
+    private final Pattern nonProxyHostsPattern;
+
+    public DeferredHttpProxyProvider(AuthorizationChallengeHandler challengeHandler,
+        AtomicReference<ChallengeHolder> proxyChallengeHolderReference, ProxyOptions proxyOptions) {
+        this.challengeHandler = challengeHandler;
+        this.proxyChallengeHolderReference = proxyChallengeHolderReference;
+        this.proxyAddress = proxyOptions.getAddress();
+        this.username = proxyOptions.getUsername();
+        this.password = proxyOptions.getPassword();
+        this.nonProxyHosts = proxyOptions.getNonProxyHosts();
+
+        this.nonProxyHostsPattern = (nonProxyHosts == null)
+            ? null
+            : Pattern.compile(nonProxyHosts, Pattern.CASE_INSENSITIVE);
+    }
+
+    @Override
+    public BiConsumer<ConnectionObserver, Channel> apply(Bootstrap bootstrap) {
+        return ((connectionObserver, channel) -> {
+            if (shouldApplyProxy(bootstrap.config().remoteAddress())) {
+                channel.pipeline()
+                    .addFirst(NettyPipeline.ProxyHandler, new HttpProxyHandler(proxyAddress, challengeHandler,
+                        proxyChallengeHolderReference))
+                    .addLast("azure.proxy.exceptionHandler", new HttpProxyExceptionHandler());
+            }
+        });
+    }
+
+    private boolean shouldApplyProxy(SocketAddress socketAddress) {
+        if (nonProxyHostsPattern == null) {
+            return true;
+        }
+
+        if (!(socketAddress instanceof  InetSocketAddress)) {
+            return true;
+        }
+
+        InetSocketAddress inetSocketAddress = (InetSocketAddress) socketAddress;
+
+        return !nonProxyHostsPattern.matcher(inetSocketAddress.getHostName()).matches();
```

Comment:
Do we need to consider (now or later) extending the pattern-matching algorithm to cover standard java networking properties formats? Reactor-netty claims to cover those cases [here](https://github.com/reactor/reactor-netty/blob/a07a8b95488ff2960ea551f4f205d3c7e6277734/src/main/java/reactor/netty/transport/ProxyProvider.java#L323).

Reply:
There is more work that is required around non proxy hosts, should roll that into this issue: https://github.com/Azure/azure-sdk-for-java/issues/7795